Improved assigning of "last-activity" and "login_date"

annando opened the pull request:
The user fields "last-activity" and "login_date" are assigned during login and the API access. Historically there had been two separate functions for that. This is now unified. Also we now always assign "login_date" only when a user is active directly on the web interface, while any API activities will solely update the "last-activity" field. Also "login_date" is updated once per hour.

These changes make these values more accurate and reliable.

annando committed:
Improved assigning of "last-activity" and "login_date"

MrPetovan commented on src/Model/User.php:
```diff
@@ -826,27 +826,26 @@ public static function getAuthenticationInfo($user_info)
 	{
-		if (!$uid) {
+		$current_day = DateTimeFormat::utcNow('Y-m-d');
+		if (($user['last-activity'] == $current_day) && (!$refresh_login || DateTimeFormat::utc($user['login_date'], 'z-H') == date('z-H'))) {
```
Why not using `utcNow` as well for the comparison? `date()` would use the local timezone, which varies depending on the server setting/location.

annando replied:
Changed.

annando committed:
Compare with the utc value

annando commented:
Why is this test failing?

MrPetovan commented:
From the log
```
1) Friendica\Test\src\Module\Api\Twitter\Users\ShowTest::testApiUsersShow
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'DFRN'
+'unkn'

tests/src/Module/Api/Twitter/Users/ShowTest.php:46
```

MrPetovan commented:
I was able to track the problem. With your changes `Authentication::setForUser` calls `User::updateLastActivity` which calls `User::update` which calls `Contact::updateSelfFromUserID` which overwrites some fields of the fixture data which leads to the discrepancy and the test failure.

MrPetovan commented:
With your changes the fixture data needs more massaging, at least:
- Add expected locality data to profile record (id = 1)
- Updating the name of the public contact (id = 48) to match the name of the test user (id = 42)

MrPetovan committed:
Normalize API fixture data

- Change public contact name to the related local user name
- Add location data to profile record that is used to update self and public contact during Auth

MrPetovan commented:
Here you go: https://github.com/annando/friendica/pull/10

annando committed:
Merge pull request #10 from MrPetovan/bug/last-activity-api-fixture-fix

Normalize API fixture data

annando commented:
@MrPetovan Done
